File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-0547.cpp
/// Disjoint set class
class DS {
    vector<int> sets;
    int numSets;
public:
    DS(int n){sets.resize(n, -1); numSets=n;}
    
    // Find set with path compression
    int find(int u){
        if (sets[u] < 0) return u;
        
        int s = find(sets[u]);
        sets[u] = s;
        return s;
    } // end-findSet
    
    void join(int u, int v){
        if (sets[u] >= 0) u = find(u);
        if (sets[v] >= 0) v = find(v);
        if (u == v) return;
        
        numSets--;
        if (sets[u] < sets[v]){
            sets[u] += sets[v];
            sets[v] = u;
        } else {
            sets[v] += sets[u];
            sets[u] = v;
        } //end-else
    } // end-join
    
    // Assume that "u" is a valid set head
    int size(int u){return -sets[u];}    
    int numberOfSets(){return numSets;}
}; // end-DS

class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();

        DS ds(n);
        for (int i=0; i<n; i++){
            for (int j=i+1; j<n; j++){
                if (isConnected[i][j]){
                    ds.join(i, j);
                } // end-if            
            } // end-for
        } // end-for
        
        return ds.numberOfSets();
    } // end-findCircleNum
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/22-UnionFind/LeetCode/LeetCode-0547.py
# Implement a Union/Find DS
# Find set with path compression
class DS:
    def __init__(self, n):
        self.sets = [-1]*n

    def find(self, u):
        if (self.sets[u] < 0): return u
        
        s = self.find(self.sets[u])
        self.sets[u] = s
        return s
    
    def join(self, u, v):
        if u == v: return
        if (self.sets[u] < self.sets[v]):
            self.sets[u] += self.sets[v]
            self.sets[v] = u
        else:
            self.sets[v] += self.sets[u]
            self.sets[u] = v

    # Assumes that "u" is a valid set head
    def size(self, u):
        return -self.sets[u]

    def noSets(self):
        count = 0
        for i in range(len(self.sets)):
            if self.sets[i] < 0:
                count += 1        
        return count

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)

        ds = DS(n)
        for i in range(n):
            for j in range(i+1, n):
                if isConnected[i][j]:
                    set1 = ds.find(i)
                    set2 = ds.find(j)
                    ds.join(set1, set2)
        return ds.noSets()
